### PR TITLE
Dynamic rules

### PR DESCRIPTION
**Pull request by ux-AylonS**

### CODE REVIEW

File: src/doc-editors/DynamicRulesEditor.vue [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        this.emitMatchDoc()
+      },
+    },
+  },
+  emits: ['update:selectedDoc', 'update:selectedDocMatchingGlobalFilter'],
+  methods: {
+    emitDocUpdate() {
+      this.$emit('update:selectedDoc', this.localDoc)
+    },
+    emitMatchDoc() {
```
Comment (Aviv-Galmidi) [inline src/doc-editors/DynamicRulesEditor.vue:291]: Let's add the `update` to the name

Perhaps:

```suggestion
    emitDocMatchUpdate() {
```

File: src/doc-editors/DynamicRulesEditor.vue [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+      this.$emit('update:selectedDocMatchingGlobalFilter', this.localGlobalFilterDoc)
+    },
+    // emitGoToRoute(url: string) {
```
Comment (Aviv-Galmidi) [inline src/doc-editors/DynamicRulesEditor.vue:294]: Remove unused code

File: src/doc-editors/DynamicRulesEditor.vue [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    //   this.$emit('go-to-route', url)
+    // },
+    emitToBoth() {
```
Comment (Aviv-Galmidi) [inline src/doc-editors/DynamicRulesEditor.vue:297]: Let's have a more meaningful name

Perhaps:

```suggestion
    emitDocAndDocMatchUpdate() {
```

File: src/doc-editors/DynamicRulesEditor.vue [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        methodName: 'GET',
+        url: `configs/${this.selectedBranch}/d/actions/`,
+        config: {headers: {'x-fields': 'id, name'}},
+      }).then((response: AxiosResponse<CustomResponse[]>) => {
+        this.customResponseNames = _.sortBy(_.map(response.data, (entity) => {
+          return [entity.id, entity.name]
+        }), (e) => {
+          return e[1]
+        })
+        console.log('this.customResponseNames', this.customResponseNames)
```
Comment (Aviv-Galmidi) [inline src/doc-editors/DynamicRulesEditor.vue:335]: Let's remove this console log

File: src/doc-editors/GlobalFilterListEditor.vue [hunk 5]
```diff
@@ -219,11 +226,19 @@ export default defineComponent({
       return `${sectionsLength} ${sectionsCounter}\t|\t${this.localDocTotalEntries} ${entriesCounter}`
     },
 
-    readonly(): boolean {
+    reblazeManaged(): boolean {
       return this.localDoc.source === 'reblaze-managed'
     },
+    dynamicRuleManaged(): boolean {
+      return this.selectedDoc.id ? this.selectedDoc.id.startsWith('dr_') : false
+      // return this.localDoc.id.startsWith('dr_')
```
Comment (Aviv-Galmidi) [inline src/doc-editors/GlobalFilterListEditor.vue:234]: Remove unused code

File: src/doc-editors/GlobalFilterListEditor.vue [hunk 6]
```diff
@@ -239,6 +254,9 @@ export default defineComponent({
         return ''
       },
       set: function(tags: string): void {
+        if ((this.localDoc.id).substr(0, 3)=='dr_') {
```
Comment (Aviv-Galmidi) [inline src/doc-editors/GlobalFilterListEditor.vue:257]:
```suggestion
        if (this.localDoc.id.startsWith('dr_')) {
```

File: src/doc-editors/SecurityPoliciesEditor.vue [hunk 7]
```diff
@@ -281,71 +281,91 @@
...
-                                  <select v-model="mapEntry.acl_profile"
-                                          @change="emitDocUpdate"
-                                          data-qa="acl-dropdown"
-                                          class="current-entry-acl-selection"
-                                          title="ACL profile">
-                                    <option v-for="acl in aclProfileNames" :value="acl[0]" :key="acl[0]">
-                                      {{ acl[1] }}
-                                    </option>
-                                  </select>
+                            <div class="column is-4">
```
Comment (Aviv-Galmidi) [inline src/doc-editors/SecurityPoliciesEditor.vue:284]: What was changed in this file? Added by mistake?

File: src/doc-editors/__tests__/RateLimitsEditor.spec.ts [hunk 8]
```diff
@@ -412,11 +412,11 @@ describe('RateLimitsEditor.vue', () => {
       const newIncludeEntryButton = filterColumn.findAll('.add-new-filter-entry-button').at(0)
       // add first
       await newIncludeEntryButton.trigger('click')
-      const firstTagAutocompleteInput = filterColumn.findComponent(TagAutocompleteInput)
-      firstTagAutocompleteInput.vm.$emit('tag-submitted', newTag)
+      const includeTagAutocompleteInput = filterColumn.findAllComponents(TagAutocompleteInput).at(0)
```
Comment (Aviv-Galmidi) [inline src/doc-editors/__tests__/RateLimitsEditor.spec.ts:415]: `filterColumn.findAllComponents(TagAutocompleteInput).at(0)` is the same as `filterColumn.findComponent(TagAutocompleteInput)`

File: src/doc-editors/__tests__/RateLimitsEditor.spec.ts [hunk 9]
```diff
@@ -397,7 +397,7 @@ describe('RateLimitsEditor.vue', () => {
         }
         return Promise.resolve()
       })
-      filterColumn = wrapper.find('.filter-column')
+      filterColumn = wrapper.find('.include-filter-column')
```
Comment (Aviv-Galmidi) [inline src/doc-editors/__tests__/RateLimitsEditor.spec.ts:400]: We're testing both include and exclude in this describe, why did you need to change this?

File: src/types/index.d.ts [hunk 10]
```diff
@@ -172,6 +179,19 @@ declare module CuriefenseClient {
...
+    id: string
+    name: string
+    active: boolean
+    description: string
+    timeframe: number
+    threshold: number
+    exclude: string[]
+    include: string[]
+    ttl: number,
+    target: 'remote_addr' | 'organization' | 'cookie' | 'geoip_city_country_name' | 'planet' | 'request_headers' | 'request_body'
```
Comment (Aviv-Galmidi) [inline src/types/index.d.ts:193]:
```suggestion
    target: DynamicRuleTargetOptionType
```

File: src/types/index.d.ts [hunk 11]
```diff
@@ -375,5 +395,15 @@ declare module CuriefenseClient {
 
   // Git - END
 
+  type Quarantined = {
+    _id: string
```
Comment (Aviv-Galmidi) [inline src/types/index.d.ts:399]: I think you've changed this to `id` on get, so the type should be `id` as well

File: src/views/DocumentEditor.vue [hunk 12]
```diff
@@ -131,10 +132,10 @@
 
                 <p class="control">
                   <button class="button is-small save-document-button"
-                          :class="{'is-loading': isSaveLoading}"
+                          :class="isSaveLoading ? 'is-loading' : 'isSaveLoading'"
```
Comment (Aviv-Galmidi) [inline src/views/DocumentEditor.vue:135]: That doesn't seem right, what was the problem with the previous code?
Currently in the new code if `isSaveLoading` is true we'll have the `is-loading` class which is correct, but if `isSaveLoading` is false we'll have `isSaveLoading` class which is not a valid class

File: src/views/DocumentEditor.vue [hunk 13]
```diff
@@ -292,25 +300,60 @@ export default defineComponent({
...
+      },
+    },
+  },
   computed: {
+    isReblazeDocument(): boolean {
+      return Object.keys(this.reblazeComponentsMap).includes(this.selectedDocType)
+    },
+
+    dynamicRuleManaged(): boolean {
+      // if (this.selectedDocType === 'globalfilters') {
```
Comment (Aviv-Galmidi) [inline src/views/DocumentEditor.vue:341]: remove unused code

File: src/views/DocumentEditor.vue [hunk 14]
```diff
@@ -399,9 +444,10 @@ export default defineComponent({
         this.selectedDocID = this.docIdNames?.[0]?.[0]
       }
       this.isDocumentInvalid = false
+
       await this.loadSelectedDocData()
       this.addMissingDefaultsToDoc()
-      await this.goToRoute()
+      // await this.goToRoute()
```
Comment (Aviv-Galmidi) [inline src/views/DocumentEditor.vue:450]: Why did you remove this?

File: src/views/DocumentEditor.vue [hunk 15]
```diff
@@ -510,9 +563,9 @@ export default defineComponent({
         this.docs = response?.data || []
         this.isDownloadLoading = false
       })
-
       this.updateDocIdNames()
       if (this.docIdNames && this.docIdNames.length && this.docIdNames[0].length) {
+        // did we move to use id only
```
Comment (Aviv-Galmidi) [inline src/views/DocumentEditor.vue:568]: What do you mean?

File: src/views/DocumentEditor.vue [hunk 16]
```diff
@@ -610,27 +667,44 @@ export default defineComponent({
       this.isNewLoading = true
       if (!docToAdd) {
         docToAdd = this.newDoc()
+        if (this.selectedDocType === 'dynamic-rules') {
```
Comment (Aviv-Galmidi) [inline src/views/DocumentEditor.vue:670]: Let's move this to DatasetsUtils - look at how we've done it for items like Routing Profiles

File: src/views/DocumentEditor.vue [hunk 17]
```diff
@@ -710,6 +798,7 @@ export default defineComponent({
       const referencedContentFilter: string[] = []
       const referencedLimit: string[] = []
       const referencedCloudFunctions: string[] = []
+      const referencedDynamicRules: string[] = []
```
Comment (Aviv-Galmidi) [inline src/views/DocumentEditor.vue:801]: Are Dynamic Rules referenced? Shouldn't need this as far as I know

File: src/views/__tests__/DocumentList.spec.ts [hunk 18]
```diff
@@ -1490,7 +1490,7 @@ describe('DocumentList.vue', () => {
       'securitypolicies',
       'contentfilterprofiles',
       'contentfilterrules',
-      'cloudfunctions',
+      'cloud-functions',
       'actions',
```
Comment (Aviv-Galmidi) [inline src/views/__tests__/DocumentList.spec.ts:1494]: add dynamic rules?

**Review by Aviv-Galmidi — CHANGES_REQUESTED** [review]

File: src/assets/DatasetsUtils.ts [hunk 19]
```diff
@@ -87,6 +92,16 @@ const limitOptionsTypes = {
   'attrs': 'Attribute',
 }
 
+const dynamicRuleTargets = {
```
[the remark this answers is not in the document]
Comment (ux-AylonS) [reply src/assets/DatasetsUtils.ts:95]: ONLY in LimitOption.vue
moved there

File: src/views/DocumentEditor.vue (hunk 14 above)
Comment (ux-AylonS) [reply src/views/DocumentEditor.vue:450]: goToRoute is await for setSelectedDataFromRouteParams
and setSelectedDataFromRouteParams awaits for goToRoute

seems like a loop

File: src/views/DocumentEditor.vue (hunk 15 above)
Comment (ux-AylonS) [reply src/views/DocumentEditor.vue:568]: Yes, this was a remark for you.
as I was not fully understand the use of doIdNames.
I removed the remark for now.
thanks

File: src/views/DocumentEditor.vue (hunk 16 above)
Comment (ux-AylonS) [reply src/views/DocumentEditor.vue:670]: made a function in DatasetUtils
function addIdtoDocName(name: string, id: string) {
  return name + ' ' + id
}

File: src/doc-editors/SecurityPoliciesEditor.vue (hunk 7 above)
Comment (ux-AylonS) [reply src/doc-editors/SecurityPoliciesEditor.vue:284]: Not sure what you are asking, without is-4 the input spread to all the width of the row
I removed it now, if that was your intention

File: src/doc-editors/__tests__/RateLimitsEditor.spec.ts (hunk 8 above)
Comment (ux-AylonS) [reply src/doc-editors/__tests__/RateLimitsEditor.spec.ts:415]: yes, findComponent find the first one,
I thought it is more specific as there are 3 uses of this component.
I changed it back in all tests

File: src/types/index.d.ts (hunk 11 above)
Comment (ux-AylonS) [reply src/types/index.d.ts:399]: right

File: src/views/DocumentEditor.vue (hunk 16 above)
Comment (Aviv-Galmidi) [reply src/views/DocumentEditor.vue:668]: I meant that we can add it to the Factory function which creates this data, similar to how we did in for Routing Profiles and the other new items

Code for example:
https://github.com/reblaze/ui/blob/main/src/assets/DatasetsUtils.ts#L328

File: src/views/DocumentEditor.vue (hunk 14 above)
Comment (Aviv-Galmidi) [reply src/views/DocumentEditor.vue:449]: It is a loop, but not an infinite loop
There's an `if` check in goToRoute

1. Route params are loaded from the route (Branch, DocType, DocId)
2. Attempt to load data using these params
2a. if all params are valid, load data and `if` check iniside goToRoute would fail
2b. if not all params are valid, reset them to the default params, and change the route to the correct one

**Review by Aviv-Galmidi — APPROVED** [review]
